Visitor can see a list of Recipes

shsteindorsson opened the pull request:
Visitor can see a list of Recipes
========================
[PT-link](https://www.pivotaltracker.com/story/show/169667218)

## PR content
In this PR we are adding a feature where the API returns a list of all recipes.

## This PR contains:
- A unit test
- An `index` action for Recipes controller
- A Recipe serializer

Commit:
Add request spec for listing articles (index)

yasmineezequiel committed:
Generates recipes controller and adds index action.

Commit:
Turn off scaffolding of extra files

Commit:
Add V1 namespace to router

Remove error from request spec.

Commit:
Add v1 namespace folder to controllers

Add headers to request spec

tochman commented on spec/requests/v1/visitor_can_see_list_of_recipes_spec.rb:
```diff
@@ -0,0 +1,45 @@
+    let(:headers) {{ HTTP_ACCEPT: "application/json" }
+    before do
+      get '/v1/recipes'
+      headers: headers
```
You need to make the headers part of the get request. Use a comma

beccaburns replied:
change articles to recipes :)

tochman commented on spec/requests/v1/visitor_can_see_list_of_recipes_spec.rb:
```diff
@@ -0,0 +1,45 @@
+    end
+
+    it 'API returns 2 recipes' do
+      expect(response_json.count).to eq 2
```
The question we need to consider is what the response should look like. I always advocate wrapping the collection in a key that holds the resource name. I know that you have done it differently in other projects, but I would like us to discuss this. 
```
{ "recipes" [

  ]
}
```

Commit:
Update language in request spec

tochman commented on spec/requests/v1/visitor_can_see_list_of_recipes_spec.rb:
```diff
@@ -0,0 +1,45 @@
+  describe 'Error if database contains no recipes' do 
+    before do
+      get '/v1/articles'
+      headers: headers
```
same here. the headers are not part of the get request

beccaburns replied:
articles to recipes :)

tochman commented on spec/requests/v1/visitor_can_see_list_of_recipes_spec.rb:
```diff
@@ -0,0 +1,45 @@
+    end
+
+    it 'returns error message' do
+      expect(response_json["message"]).to eq "There are no recipes here"
```
This is not an error message ;-)

tochman reviewed (commented):
I can't really see how the tests are supposed to pass on this one without a model and factory....

Commit:
Merge branch 'development' of github.com:CraftAcademy/foodhub-api into list_recipes

Fix merge conflicts.
Update error responses.
Update list recipes response to wrap in a hash.
Fix typos.

oliverochman commented on coverage/.resultset.json:
```diff
@@ -0,0 +1,81 @@
+{
```
you need to add coverage to the `.gitignore` and remove the git cache

http://www.codeblocq.com/2016/01/Untrack-files-already-added-to-git-repository-based-on-gitignore/

oliverochman commented on app/controllers/v1/recipes_controller.rb:
```diff
@@ -1,6 +1,17 @@
+    if recipes.empty?
+      render json: { message: 'There are no recipes here' }
+    else 
+      render json: { "recipes": recipes }
```
please serialize the response

Commit:
.gitignore fix

tochman commented on app/controllers/v1/recipes_controller.rb:
```diff
@@ -0,0 +1,12 @@
+    recipes = Recipe.all
+
+    if recipes.empty?
+      render_message('There are no recipes here')
```
Please just render a response. W spoke about this at the meeting.

Commit:
Generate serializer for recipe

tochman commented on app/controllers/v1/recipes_controller.rb:
```diff
@@ -1,6 +1,17 @@
+
+    if recipes.empty?
+      render json: { message: 'There are no recipes here' }
+    else
```
I will again suggest that you just render an empty collection and deal with it on the client-side. The way you do it now will still require you to add an error handler when you receive the response.

Commit:
Remove binding.pry

Commit:
Change controller response for no recipes

Controller responds with same hash format as positive response.
Update request spec to expect new format.

Commit:
Refactor index action & remove extra lines
